python/ql/test/experimental/dataflow/coverage/test.py

```diff
@@ -516,6 +516,110 @@ def test_swap():
     SINK(b)
 
 
+@expects(2)
+def test_unpacking_assignment():
+    t = (SOURCE, NONSOURCE)
+    a, b = t
+    SINK(a)
+    SINK_F(b)
+
+
+@expects(3)
+def test_nested_unpacking_assignment():
+    t = (SOURCE, (NONSOURCE, SOURCE))
+    a, (b, c) = t
+    SINK(a)
+    SINK_F(b)
+    SINK(c)
+
+
+@expects(2)
+def test_deeply_nested_unpacking_assignment():
+    t = [[[[SOURCE]]], NONSOURCE]
+    [[[a]]], b = t
+    SINK(a)
+    SINK_F(b)
+
+
+@expects(4)
+def test_iterated_unpacking_assignment():
+    t = (SOURCE, SOURCE, NONSOURCE)
+    a, *b, c = t
+    SINK(a)
+    SINK_F(b)
+    SINK(b[0])
+    SINK_F(c)
+
+
+@expects(15)
+def test_unpacking_assignment_conversion():
+    ll = [[SOURCE, NONSOURCE, SOURCE], [SOURCE], [NONSOURCE]]
+
+    # tuple
+    ((a1, a2, a3), b, c) = ll
+    SINK(a1)
+    SINK_F(a2)  # We expect an FP as all elements are tainted
```

Comment:
This bothers me a bit. It feels slightly wrong to use "any element can be tainted" when we actually have enough information to reconstruct which elements are in fact tainted.

Perhaps we could do something along these lines:

- Track list indices precisely for the cases where are able, _but_
- have this gracefully devolve into "some element is tainted" when we perform operations that may move things around (so something like `list.append` is probably fine, as it only adds to the end of the list. `list.sort` would go from specific to generic taint, etc.)

This should probably not happen in this PR, though.

Reply:
I would be happy to try this out at some point. Note, though, that we are already more precise than C# where tuple elements are modelled as `ElementContent`, so we may or may not need al this precision. I agree that introducing precise element information and having a graceful mutation to imprecise information is interesting, not only for lists. But I do think it is out of scope for this PR.